x/dsmr/node_test.go

```diff
@@ -1526,6 +1516,79 @@ func newTestNodes(t *testing.T, n int) []*Node[dsmrtest.Tx] {
 	return result
 }
 
+type testChainState struct {
+	validators []Validator
+	quorumNum  uint64
+	quorumDen  uint64
+}
+
+func (*testChainState) GetNetworkID() uint32 {
+	return networkID
+}
+
+func (*testChainState) GetSubnetID() ids.ID {
+	return ids.Empty
+}
+
+func (*testChainState) GetChainID() ids.ID {
+	return chainID
+}
+
+func (t *testChainState) GetQuorumNum() uint64 {
+	return t.quorumNum
+}
+
+func (t *testChainState) GetQuorumDen() uint64 {
+	return t.quorumDen
+}
+
+func (t *testChainState) getValidatorSet() map[ids.NodeID]*validators.GetValidatorOutput {
+	result := make(map[ids.NodeID]*validators.GetValidatorOutput, len(t.validators))
+	for _, v := range t.validators {
+		result[v.NodeID] = &validators.GetValidatorOutput{
+			NodeID:    v.NodeID,
+			PublicKey: v.PublicKey,
+			Weight:    v.Weight,
+		}
+	}
+	return result
+}
+
+func (t *testChainState) GetCanonicalValidatorSet(ctx context.Context) (warp.CanonicalValidatorSet, error) {
+	pChain := &validatorstest.State{
+		GetSubnetIDF: func(context.Context, ids.ID) (ids.ID, error) {
+			return t.GetSubnetID(), nil
+		},
+		GetValidatorSetF: func(
+			context.Context,
+			uint64,
+			ids.ID,
+		) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
+			return t.getValidatorSet(), nil
+		},
+	}
+
+	canonicalValidators, err := warp.GetCanonicalValidatorSetFromSubnetID(
```

Comment:
if we need this, can we include it in the `testChainState` type rather than constructing it on demand in this function?

Reply:
sure. it would require a constructor, but no prob.